File: sdks/php/tests/TopicsTest.php
<?php

use Kibamail\Collection;
use Kibamail\Topic;

it('can create a topic', function () {
    $client = createClient();

    $result = $client->topics->create([
        'name' => 'Product Updates',
        'description' => 'Latest product news',
        'visibility' => 'PUBLIC',
        'defaultOptIn' => true,
    ]);

    expect($result)->toBeInstanceOf(Topic::class)
        ->and($result->id)->not()->toBeNull();
});

it('can list topics', function () {
    $client = createClient();

    $result = $client->topics->list();

    expect($result)->toBeInstanceOf(Collection::class)
        ->and($result->data())->toBeArray();
});

it('can get a topic', function () {
    $client = createClient();

    $topic = $client->topics->create([
        'name' => 'Newsletter',
    ]);

    $result = $client->topics->get($topic->id);

    expect($result)->toBeInstanceOf(Topic::class)
        ->and($result->id)->toBe($topic->id);
});

it('can update a topic', function () {
    $client = createClient();

    $topic = $client->topics->create([
        'name' => 'Original Name',
    ]);

    $result = $client->topics->update($topic->id, [
        'name' => 'Updated Name',
    ]);

    expect($result)->toBeInstanceOf(Topic::class);
});

it('can delete a topic', function () {
    $client = createClient();

    $topic = $client->topics->create([
        'name' => 'To Delete',
    ]);

    $result = $client->topics->delete($topic->id);

    expect($result)->toBeInstanceOf(Topic::class);
});

File: sdks/go/topics_test.go
package kibamail

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestTopicsCreate(t *testing.T) {
	client := setupTest()

	t.Run("should create a topic with basic information", func(t *testing.T) {
		result, err := client.Topics.Create(&CreateTopicRequest{
			Name:        "Product Updates",
			Description: "Latest features and improvements",
			Visibility:  "PUBLIC",
		})

		require.NoError(t, err)
		require.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})

	t.Run("should create a topic with default opt-in", func(t *testing.T) {
		result, err := client.Topics.Create(&CreateTopicRequest{
			Name:         "Newsletter",
			Description:  "Weekly newsletter",
			Visibility:   "PUBLIC",
			DefaultOptIn: true,
		})

		require.NoError(t, err)
		require.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})
}

func TestTopicsList(t *testing.T) {
	client := setupTest()

	t.Run("should list topics with default pagination", func(t *testing.T) {
		result, err := client.Topics.List(nil)

		require.NoError(t, err)
		require.NotNil(t, result)
		assert.NotNil(t, result.Data)
	})

	t.Run("should list topics with custom limit", func(t *testing.T) {
		limit := 50
		result, err := client.Topics.List(&ListOptions{Limit: &limit})

		require.NoError(t, err)
		require.NotNil(t, result)
		assert.NotNil(t, result.Data)
	})
}

func TestTopicsGet(t *testing.T) {
	client := setupTest()

	t.Run("should retrieve a topic by ID", func(t *testing.T) {
		topicId := "topic_test_12345"
		result, err := client.Topics.Get(topicId)

		require.NoError(t, err)
		require.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})
}

func TestTopicsUpdate(t *testing.T) {
	client := setupTest()

	t.Run("should update a topic's basic information", func(t *testing.T) {
		topicId := "topic_test_12345"
		result, err := client.Topics.Update(topicId, &UpdateTopicRequest{
			Name:        "Updated Topic",
			Description: "Updated description",
		})

		require.NoError(t, err)
		require.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})

	t.Run("should update a topic's visibility", func(t *testing.T) {
		topicId := "topic_test_12345"
		result, err := client.Topics.Update(topicId, &UpdateTopicRequest{
			Visibility: "PRIVATE",
		})

		require.NoError(t, err)
		require.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})
}

func TestTopicsDelete(t *testing.T) {
	client := setupTest()

	t.Run("should delete a topic by ID", func(t *testing.T) {
		topicId := "topic_test_12345"
		err := client.Topics.Delete(topicId)

		require.NoError(t, err)
	})
}
